File: 102-Binary-Tree-Level-Order-Traversal.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
from collections import deque
class Solution:
    def levelOrder(self, root: Optional[TreeNode]) -> List[List[int]]:
        if not root:
            return []

        queue = [root]
        visited = []

        while queue:
            level_size = len(queue)
            level = []
            for i in range(level_size):
                tempRoot = queue.pop(0)
                level.append(tempRoot.val)
                if tempRoot.left:
                    queue.append(tempRoot.left)
                if tempRoot.right:
                    queue.append(tempRoot.right)
            visited.append(level)
        return visited

File: 102-Binary-Tree-Level-Order-Traversal.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public List<List<Integer>> levelOrder(TreeNode root) {
        List<List<Integer>> levels = new ArrayList<>();
        if(root == null)    return levels;
        Queue<TreeNode> nodeLevel = new LinkedList<>();

        nodeLevel.offer(root);
        while(!nodeLevel.isEmpty()){
            int N = nodeLevel.size();
            List<Integer> level = new ArrayList<>();
            for(int i = 0; i < N; i++){
                TreeNode dummy = nodeLevel.poll();
                if(dummy.left != null){
                    nodeLevel.offer(dummy.left);
                }
                if(dummy.right != null){
                    nodeLevel.offer(dummy.right);
                }
                level.add(dummy.val);
            }
            levels.add(level);
        }
        return levels;
    }
}